File: articles/calculus-for-systems-modeling/articles/finite-difference-methods/cpp/finite_difference_audit.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>
int main(){ int n=61, steps=120; double dx=1.0, dt=0.2, ratio=0.08*0.2; std::vector<double> field(n,0.0); field[n/2]=1.0; std::cout<<"step,time,center_value,total_mass,max_value,left_boundary,right_boundary,diffusion_ratio\n"; for(int s=0;s<=steps;s++){ double total=std::accumulate(field.begin(), field.end(), 0.0)*dx; double maxv=*std::max_element(field.begin(), field.end()); std::cout<<s<<","<<s*dt<<","<<field[n/2]<<","<<total<<","<<maxv<<","<<field.front()<<","<<field.back()<<","<<ratio<<"\n"; auto updated=field; for(int i=1;i<n-1;i++) updated[i]=field[i]+ratio*(field[i+1]-2*field[i]+field[i-1]); updated.front()=0; updated.back()=0; field=updated; } }

File: articles/calculus-for-systems-modeling/articles/finite-difference-methods/fortran/finite_difference_audit.f90
program finite_difference_audit
implicit none
integer, parameter :: n=61, steps=120
integer :: s,i
real(8) :: field(n), updated(n), ratio, dx, dt, total_mass
dx=1.0d0; dt=0.2d0; ratio=0.08d0*dt/(dx*dx); field=0.0d0; field(31)=1.0d0
print '(A)', 'step time center_value total_mass diffusion_ratio'
do s=0,steps
  total_mass=sum(field)*dx
  print '(I6,4F14.6)', s, dble(s)*dt, field(31), total_mass, ratio
  updated=field
  do i=2,n-1
    updated(i)=field(i)+ratio*(field(i+1)-2.0d0*field(i)+field(i-1))
  end do
  updated(1)=0.0d0; updated(n)=0.0d0; field=updated
end do
end program
